Connection Postgres

ERROR: FATAL: too many connections for role "mssbzhocnhzlzu" 
Jun 07 00:34:00 takemyplace app/web.1:  [qtp1091825220-46] ERROR spark.webserver.MatcherFilter -  
Jun 07 00:34:00 takemyplace app/web.1:  org.hibernate.exception.GenericJDBCException: Error calling Driver#connect 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:47) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:111) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.hibernate.engine.jdbc.connections.internal.BasicConnectionCreator.convertSqlException(BasicConnectionCreator.java:118) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.hibernate.engine.jdbc.connections.internal.DriverConnectionCreator.makeConnection(DriverConnectionCreator.java:41) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.hibernate.engine.jdbc.connections.internal.BasicConnectionCreator.createConnection(BasicConnectionCreator.java:58) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.hibernate.engine.jdbc.connections.internal.DriverManagerConnectionProviderImpl.getConnection(DriverManagerConnectionProviderImpl.java:189) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.hibernate.internal.AbstractSessionImpl$NonContextualJdbcConnectionAccess.obtainConnection(AbstractSessionImpl.java:386) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl.acquireConnectionIfNeeded(LogicalConnectionManagedImpl.java:84) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl.getPhysicalConnection(LogicalConnectionManagedImpl.java:109) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.hibernate.engine.jdbc.internal.StatementPreparerImpl.connection(StatementPreparerImpl.java:47) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.hibernate.engine.jdbc.internal.StatementPreparerImpl$5.doPrepare(StatementPreparerImpl.java:146) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.hibernate.engine.jdbc.internal.StatementPreparerImpl$StatementPreparationTemplate.prepareStatement(StatementPreparerImpl.java:172) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.hibernate.engine.jdbc.internal.StatementPreparerImpl.prepareQueryStatement(StatementPreparerImpl.java:148) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.hibernate.loader.Loader.prepareQueryStatement(Loader.java:1927) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:1896) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:1874) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.hibernate.loader.Loader.doQuery(Loader.java:919) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.hibernate.loader.Loader.doQueryAndInitializeNonLazyCollections(Loader.java:336) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.hibernate.loader.Loader.doList(Loader.java:2610) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.hibernate.loader.Loader.doList(Loader.java:2593) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.hibernate.loader.Loader.listIgnoreQueryCache(Loader.java:2422) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.hibernate.loader.Loader.list(Loader.java:2417) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.hibernate.loader.criteria.CriteriaLoader.list(CriteriaLoader.java:109) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.hibernate.internal.SessionImpl.list(SessionImpl.java:1787) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.hibernate.internal.CriteriaImpl.list(CriteriaImpl.java:363) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.hibernate.internal.CriteriaImpl.uniqueResult(CriteriaImpl.java:385) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.friends.app.dao.PlaceDao.findPlaceByCriterions(PlaceDao.java:88) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.friends.app.service.impl.PlaceServiceBean.isPlaceShared(PlaceServiceBean.java:219) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.friends.app.view.route.ShareRoute.doHandle(ShareRoute.java:42) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.friends.app.view.route.AuthenticatedRoute.handle(AuthenticatedRoute.java:26) 
Jun 07 00:34:00 takemyplace app/web.1:  	at spark.TemplateViewRouteImpl$1.handle(TemplateViewRouteImpl.java:66) 
Jun 07 00:34:00 takemyplace app/web.1:  	at spark.webserver.MatcherFilter.doFilter(MatcherFilter.java:162) 
Jun 07 00:34:00 takemyplace app/web.1:  	at spark.webserver.JettyHandler.doHandle(JettyHandler.java:61) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:189) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:52) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:119) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.eclipse.jetty.server.Server.handle(Server.java:517) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:302) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:242) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:245) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:95) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.eclipse.jetty.io.SelectChannelEndPoint$2.run(SelectChannelEndPoint.java:75) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceAndRun(ExecuteProduceConsume.java:213) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run(ExecuteProduceConsume.java:147) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:654) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:572) 
Jun 07 00:34:00 takemyplace app/web.1:  	at java.lang.Thread.run(Thread.java:745) 
Jun 07 00:34:00 takemyplace app/web.1:  Caused by: org.postgresql.util.PSQLException: FATAL: too many connections for role "mssbzhocnhzlzu" 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.postgresql.core.v3.ConnectionFactoryImpl.readStartupMessages(ConnectionFactoryImpl.java:712) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:213) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:67) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:216) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.postgresql.Driver.makeConnection(Driver.java:406) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.postgresql.Driver.connect(Driver.java:274) 
Jun 07 00:34:00 takemyplace app/web.1:  	at org.hibernate.engine.jdbc.connections.internal.DriverConnectionCreator.makeConnection(DriverConnectionCreator.java:38) 
Jun 07 00:34:00 takemyplace app/web.1:  	... 44 more

--- a/pom.xml
+++ b/pom.xml
@@ -70,6 +70,16 @@
         <artifactId>hibernate-core</artifactId>
         <version>${hibernate-version}</version>
     </dependency>
+    <dependency>
+		<groupId>org.hibernate</groupId>
+		<artifactId>hibernate-c3p0</artifactId>
+		<version>${hibernate-version}</version>
+	</dependency>
+	<dependency>
+		<groupId>com.mchange</groupId>
+		<artifactId>c3p0</artifactId>
+		<version>0.9.5.2</version>
+	</dependency>
   </dependencies>
   <build>
     <plugins>

--- a/src/main/java/org/friends/app/HibernateUtil.java
+++ b/src/main/java/org/friends/app/HibernateUtil.java
@@ -1,5 +1,7 @@
 package org.friends.app;
 
+import java.sql.Connection;
+
 import org.friends.app.model.Place;
 import org.friends.app.model.User;
 import org.hibernate.Session;
@@ -12,56 +14,67 @@ import org.slf4j.LoggerFactory;
 
 public class HibernateUtil {
 
-    private static final Logger LOGGER = LoggerFactory.getLogger(HibernateUtil.class);
-    
-    private static final SessionFactory SESSION_FACTORY = createSessionFactory();
-    private static final ThreadLocal<Session> SESSIONS = new ThreadLocal<>();
-    
-    public static Session getSession() {
-        Session session = SESSIONS.get();
-        if (session == null) {
-            session = SESSION_FACTORY.openSession();
-            SESSIONS.set(session);
-        }
-        return session;
-    }
-    
-    public static void closeSession() {
-        Session session = SESSIONS.get();
-        if (session != null) {
-            session.close();
-            SESSIONS.remove();
-        }
-    }
-    
-    private static SessionFactory createSessionFactory() {
-        try {
-            Configuration configuration = configuration();
-            configuration.setProperty(AvailableSettings.URL, org.friends.app.Configuration.dbUrl());
-			configuration.setProperty(AvailableSettings.DIALECT, org.friends.app.Configuration.dialect());
-            configuration.setProperty(AvailableSettings.USE_QUERY_CACHE, "false");
-            configuration.setProperty(AvailableSettings.SHOW_SQL, "true");
-            configuration.setProperty(AvailableSettings.CURRENT_SESSION_CONTEXT_CLASS, "thread");
-            configuration.addAnnotatedClass(Place.class);
-            configuration.addAnnotatedClass(org.friends.app.model.Session.class);
-            configuration.addAnnotatedClass(User.class);
-            StandardServiceRegistryBuilder serviceRegistryBuilder = new StandardServiceRegistryBuilder().applySettings(configuration.getProperties());
-            return configuration.buildSessionFactory(serviceRegistryBuilder.build());
-        } catch (RuntimeException ex) {
-            LOGGER.error("Failed to create session factory");
-            throw ex;
-        }
-    }
- 
-    private static Configuration configuration() {
-        String confFile = "/hibernate-local.cfg.xml";
-        if (!org.friends.app.Configuration.development()) {
-        	confFile = "/hibernate-jndi.cfg.xml";
-        }
+	private static final Logger LOGGER = LoggerFactory.getLogger(HibernateUtil.class);
 
-        Configuration configuration = new Configuration();
-        configuration.configure(confFile);
-        return configuration;
-    }
-    
+	private static final SessionFactory SESSION_FACTORY = createSessionFactory();
+	private static final ThreadLocal<Session> SESSIONS = new ThreadLocal<>();
+
+	public static Session getSession() {
+		Session session = SESSIONS.get();
+		if (session == null) {
+			session = SESSION_FACTORY.openSession();
+			SESSIONS.set(session);
+		}
+		return session;
+	}
+
+	public static void closeSession() {
+		Session session = SESSIONS.get();
+		if (session != null) {
+			session.close();
+			SESSIONS.remove();
+		}
+	}
+
+	private static SessionFactory createSessionFactory() {
+		try {
+			Configuration configuration = configuration();
+			configuration.setProperty(AvailableSettings.URL, org.friends.app.Configuration.dbUrl());
+			configuration.setProperty(AvailableSettings.DIALECT, org.friends.app.Configuration.dialect());
+			configuration.setProperty(AvailableSettings.USE_QUERY_CACHE, "false");
+			configuration.setProperty(AvailableSettings.SHOW_SQL, "true");
+			configuration.setProperty(AvailableSettings.CURRENT_SESSION_CONTEXT_CLASS, "thread");
+			configuration.setProperty(AvailableSettings.ISOLATION,
+					String.valueOf(Connection.TRANSACTION_READ_UNCOMMITTED));
+
+			configuration.setProperty("hibernate.c3p0.min_size", "10");
+			configuration.setProperty("hibernate.c3p0.max_size", "20");
+			configuration.setProperty("hibernate.c3p0.acquire_increment", "1");
+			configuration.setProperty("hibernate.c3p0.idle_test_period", "3000");
+			configuration.setProperty("hibernate.c3p0.max_statements", "50");
+			configuration.setProperty("hibernate.c3p0.timeout", "1800");
+
+			configuration.addAnnotatedClass(Place.class);
+			configuration.addAnnotatedClass(org.friends.app.model.Session.class);
+			configuration.addAnnotatedClass(User.class);
+			StandardServiceRegistryBuilder serviceRegistryBuilder = new StandardServiceRegistryBuilder()
+					.applySettings(configuration.getProperties());
+			return configuration.buildSessionFactory(serviceRegistryBuilder.build());
+		} catch (RuntimeException ex) {
+			LOGGER.error("Failed to create session factory");
+			throw ex;
+		}
+	}
+
+	private static Configuration configuration() {
+		String confFile = "/hibernate-local.cfg.xml";
+		if (!org.friends.app.Configuration.development()) {
+			confFile = "/hibernate-jndi.cfg.xml";
+		}
+
+		Configuration configuration = new Configuration();
+		configuration.configure(confFile);
+		return configuration;
+	}
+
 }

